frontend/src/components/translation/Autosave.tsx

```diff
@@ -0,0 +1,75 @@
+import { useCallback, useEffect } from "react";
+import { gql, useMutation } from "@apollo/client";
+
+import debounce from "../../utils/DebounceUtils";
+
+export type StoryLine = {
+  lineIndex: number;
+  originalContent: string;
+  translatedContent?: string;
+  storyTranslationContentId?: number;
+};
+
+const UPDATE_TRANSLATION = gql`
+  mutation updateStoryTranslationContents(
+    $storyTranslationContents: [StoryTranslationContentRequestDTO]
+  ) {
+    updateStoryTranslationContents(
+      storyTranslationContents: $storyTranslationContents
+    ) {
+      story {
+        id
+      }
+    }
+  }
+`;
+
+type AutosaveProps = {
+  storylines: StoryLine[];
+  onSuccess: () => void;
+};
+
+// Inspiration from https://www.synthace.com/autosave-with-react-hooks/
+const Autosave = ({ storylines, onSuccess }: AutosaveProps) => {
+  const handleError = (errorMessage: string) => {
+    alert(errorMessage);
+  };
+
+  const [updateTranslation] = useMutation<{}>(UPDATE_TRANSLATION);
+
+  const debouncedSave = useCallback(
+    debounce(async (linesToUpdate: StoryLine[]) => {
+      const storyTranslationContents = linesToUpdate.map((line: StoryLine) => {
+        return {
+          id: line.storyTranslationContentId,
+          translationContent: line.translatedContent,
+        };
+      });
+
+      try {
+        const result = await updateTranslation({
+          variables: { storyTranslationContents },
+        });
+
+        if (result.data == null) {
+          handleError("Unable to save translation");
+        } else if (linesToUpdate.length > 0) {
```

Comment:
This condition strikes me as a little odd - I believe `linesToUpdate.length > 0` will always be true if `linesToUpdate `was passed with any length - did you mean something like `result.data.length > 0`

Reply:
Hmm don't remember why I added this, might've been a "make it work" type condition because at one point requests were firing more than expected. Will check.

Reply:
I think it's just to prevent unnecessary calls that would prompt a re-render on parent page. Don't think result.data.length is anything useful?

Changed it up so that there is a length check at the very beginning. Results with no changes if so.

Reply:
Interesting, thanks for humoring me
Happy with the current code